### PR TITLE
Add ability to skip authentication based on request headers

### PR DESCRIPTION
**Pull request by fnkr**
## Description

This adds the ability to skip authentication based on request headers. Request headers are compared to a list of regular expressions provided in the configuration. If any header matches any of the provided regular expressions, oauth2proxy will grant access.

## Motivation and Context

We're using an internal fork of the bitly/oauth2_proxy project and like to switch to pusher/oauth2_proxy. We have several use cases for this, for example:

- `Host`: Some applications use different domains for different services. You might want to skip access control for some of them.
- `X-Real-IP`: If oauth2proxy is running behind a reverse proxy, this can be used to skip access control for certain IP ranges.
- `User-Agent`: This can be used to skip access control for bots.

## How Has This Been Tested?

- If the option is not used, there is no change in behavior.
- If the command line option `--skip-auth-header '(?i)^User-Agent:.*curl.*'` or config file option `skip_auth_header = ["(?i)^User-Agent:.*curl.*'"]` is set, the site can be accessed via curl but not via Chrome.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Hey @fnkr, thanks for raising this PR. It's currently got some conflicts with master so I'd appreciate if you could fix those up

I feel like this option would maybe be more secure/useful if you could pair these regex's with particular paths?

I'm concerned a little bit about how easy it might be for attackers to forge headers and get access to protected assets, if a user allowed bot user agents to access without authenticating, then I feel like this could very easily be used to bypass auth for all resources right?

Your example use cases mention having a reverse proxy in front of this project, have you considered moving this logic to the reverse proxy instead?

**Comment by fnkr** [comment]
Hi @JoelSpeed!

thanks for maintaining this project, it is appreciated!

> It's currently got some conflicts with master so I'd appreciate if you could fix those up

Done.

> I feel like this option would maybe be more secure/useful if you could pair these regex's with particular paths?

I don't have any use case for this right now but i'll add it if you want. How should the configuration look like?

> I'm concerned a little bit about how easy it might be for attackers to forge headers and get access to protected assets, if a user allowed bot user agents to access without authenticating, then I feel like this could very easily be used to bypass auth for all resources right?

Yes, and I wouldn't recommend to use the `.*curl.*` example _as is_. A real world use case looks like this: we use oauth2-proxy on our staging environments to prevent bots and curious visitors from snooping around - but some SEO tools still need to have access to the site. Usually basic auth doesn't work (either because oauth2-proxy returns status code 200 instead of 401 for unauthenticated requests, or because they don't support it at all), but they can be identified by their User-Agent - and the User-Agent can be configured to be something unique (long random string)).

Also I want to add that there are many ways to break security, e.g. by configuring a public identity provider without limiting access to certain groups (or [by implementing it wrong](https://github.com/bitly/oauth2_proxy/pull/312#discussion_r222286809)). Whoever configures the auth proxy should know what they're doing otherwise it will be bad. But we can add a warning if you want :)

> Your example use cases mention having a reverse proxy in front of this project, have you considered moving this logic to the reverse proxy instead?

Yes, and to be honest I'd prefer to do that. The problem is that it's not that easy and oftentimes not possible at all. For example, one of the web servers we use a lot is nginx. In this scenario we use the `auth_subrequest` directive to forward authentication to oauth2-proxy. Unfortunately, there is no way to tell nginx to skip authentication based on a header. You can't just use `if`, or `map` or anything useful in that context. The only directive that can control whether the `auth_request` is required is [satisfy](https://nginx.org/en/docs/http/ngx_http_core_module.html#satisfy) and that only helps if you can identify allowable requests by IP or basic auth. Even with OpenResty (nginx+lua) it's not possible (at least not if http2 is enabled).

File: oauthproxy.go [hunk 1]
```diff
@@ -623,7 +630,23 @@ func (p *OAuthProxy) IsValidRedirect(redirect string) bool {
...
 	isPreflightRequestAllowed := p.skipAuthPreflight && req.Method == "OPTIONS"
-	return isPreflightRequestAllowed || p.IsWhitelistedPath(req.URL.Path)
+	return isPreflightRequestAllowed || p.IsWhitelistedPath(req.URL.Path) || p.IsWhitelistedHeader(req)
+}
+
+func (p *OAuthProxy) IsWhitelistedHeader(req *http.Request) bool {
+	for _, expr := range p.compiledHeader {
+		if expr.MatchString("Host: " + req.Host) {
+			return true
+		}
```
Comment (JoelSpeed) [inline oauthproxy.go:640]: Surely this can be achieved via the normal header values?

**Comment by fnkr** [comment]
> I'd certainly feel happier about the security if it were tied to paths, ie, allow this header to access this particular path.

Ok, I'll add it. How about this: `--skip-auth-header '/admin/?(?i)^User-Agent:.*curl.*'`? (Using `?`  as separator because paths cannot contain `?`).

> This is hard, we don't yet have structured configuration but maybe that could be introduced once we do (see #532)

Structured configuration would be nice. I've left a comment in the other issue.

File: oauthproxy.go (hunk 1 above)
Comment (fnkr) [reply oauthproxy.go:640]: `req.Header` does not contain the `Host` header:

> For incoming requests, the Host header is promoted to the
> Request.Host field and removed from the Header map.

https://golang.org/pkg/net/http/#Request

Comment (JoelSpeed) [reply oauthproxy.go:640]: This makes sense, might be good to have a comment to that effect in the code as well?

**Comment by JoelSpeed** [comment]
> Ok, I'll add it. How about this: --skip-auth-header '/admin/?(?i)^User-Agent:.*curl.*'? (Using ? as separator because paths cannot contain ?).

A `?` character is valid in regex though, so I'm not sure that's the best character to use either. Other options I believe use a `|` character as the delimeter, though I'm also not a huge fan of this.

> Structured configuration would be nice. I've left a comment in the other issue.

How would you feel about waiting for the structured configuration before merging this, that we don't have to design something that we know is going to be removed soonish

**Comment by fnkr** [comment]
> A `?` character is valid in regex though, so I'm not sure that's the best character to use either. Other options I believe use a `|` character as the delimeter, though I'm also not a huge fan of this.

In my example `/admin/` is a path, not a regex (the format is `path?regex`, where `path` is the URI prefix and `regex` is compared to each header). The regex only starts after the first `?` so there is no conflict (unless you want the path also to be a regex, otherwise I would just use `strings.HasPrefix()`).

> How would you feel about waiting for the structured configuration before merging this, that we don't have to design something that we know is going to be removed soonish

We're using our fork for quite some time now, I guess we can wait a bit longer. Do you have any idea how long it will take for this to be implemented and ready?

**Comment by JoelSpeed** [comment]
> In my example /admin/ is a path, not a regex (the format is path?regex, where path is the URI prefix and regex is compared to each header). The regex only starts after the first ? so there is no conflict (unless you want the path also to be a regex, otherwise I would just use strings.HasPrefix()).

I don't think the path should be a regex, so I guess the `?` separator is ok. Have you considered `|` as an alternative?

> We're using our fork for quite some time now, I guess we can wait a bit longer. Do you have any idea how long it will take for this to be implemented and ready?

I've put a substantial amount of work into it recently, hoping to get it all done within the next couple of months. Problem there being I'm waiting on reviews now because there's a backlog of PRs that all touch similar bits of code 😅

**Comment by stefansedich** [comment]
This one would be handy for us too, we are fronting MLFlow with oauth2-proxy and while user auth works great we need to allow our backend services to authenticate too.

We use dex and our tokens are short lived so I can't just use an id-token like I had intended, but would be handy if I could have some static token and I could verify the Authorizatio header contains this token value.

**Comment by fnkr** [comment]
Can we keep this open for now?

**Comment by fnkr** [comment]
Not stale.

**Comment by JoelSpeed** [comment]
We have been talking recently about trying to add some new alpha structured config options, do you think you'd be willing to pick this up again once we have the initial alpha config options in? I can ping back here when that's ready? 

The idea basically is that if you want to have the new features you have to switch to the alpha config, kinda like a feature gate style concept

**Comment by fnkr** [comment]
Sounds good, please ping me back.

**Comment by JoelSpeed** [comment]
@fnkr So you're aware, I'm working on the Alpha config stuff in #907, I would expect the configuration for this feature to become a new structure in the options package and be a member of both `Options` and `AlphaOptions`. We would then allow users to only use the feature if they are using the `--alpha-config` flag being introduced

File: oauthproxy.go [hunk 2]
```diff
@@ -376,6 +389,44 @@ func buildRoutesAllowlist(opts *options.Options) ([]allowedRoute, error) {
 	return routes, nil
 }
 
+// buildHeadersAllowlist builds an []allowedHeader list from the SkipAuthHeaders and LegacySkipAuthHeaders options
+func buildHeadersAllowlist(opts *options.Options) ([]allowedHeader, error) {
```
Comment (fnkr) [inline oauthproxy.go:393]: codeclimate complains about too many return statements. Should I move `appendAllowedHeader()` out of here (make it non-anonymous)?

**Comment by fnkr** [comment]
Hi @JoelSpeed,

I updated the pull request, now using `alpha-config` as base branch. There are still some parts missing (mostly tests and documentation) but the feature itself is working. Maybe you can let me know what you're thinking?

Best regards

**Comment by fnkr** [comment]
Hi @JoelSpeed, was this closed intentionally?

**Comment by fnkr** [comment]
I can't change the base branch (probably because the PR is closed) and I can't reopen the PR (probably because the base branch does not exist). Can you temporarily restore the `oauth2-proxy:alpha-config` branch (there should be a "restore branch" button in the https://github.com/oauth2-proxy/oauth2-proxy/pull/907 PR). I think I can reopen this PR then and change the base branch back to master.

**Comment by JoelSpeed** [comment]
Good idea! I've restored the branch, reopened the PR and changed the base branch for you. In need of a rebase but I think we are good to carry on

**Comment by fnkr** [comment]
Still relevant.

**Comment by NickMeves** [comment]
I have strong apprehensions about supporting this functionality for security reasons.

I see tons of opportunities for people to implement this while not realizing the potential backdoors they are opening themselves up to where hackers could skip strong OAuth2 based authentication and get in by injecting the right header.

(E.g. the Bot User-Agent example you gave).

Until we have a stronger authorization engine where various factors can be intermixed to skip authorization (e.g. this user-agent, this HTTP Request, this specific route) -- skipping authentication & authorization solely based on Request Headers feels too risky.

I still even have my apprehensions for IP-based authentication skipping due to scenarios behind a proxy, it will read from the `X-Forwarded-For` header and is very spoofable (Hence we give a big scary warning with flags that would result in this security vulnerability are enabled simultaneously).

**Comment by NickMeves** [comment]
I'm going to go ahead and close this for the time being. I don't want to expose arbitrary header based allowlists in the project at the moment for the reasons I mentioned above.

Someday (v9 timeframe?) We are looking at rich authorization rules based on Open Policy Agent (https://www.openpolicyagent.org/) -- when that is in place, something like this could be combined with other factors to make a more secure allowlist.

**Pull request closed without merging**